### ISSUE TITLE
Notify only on state change

### ISSUE THREAD

**Issue by hchargois** [sensu/sensu#1402]
Hi,

My team is using sensu to replace nagios. We are newcomers to sensu. It took us many days of tedious work to get it working, but now it works, we get alerts by mail and Slack... Well, too many of them actually. We are getting alerts every 30 minutes for the very same failing checks.

Is there a way to get a sane configuration, that is, to get notified only once, when the check begins to fail, and then not until it is resolved? Nagios does this out of the box. That's the obvious default.

The [documentation] says "A Sensu Event is created when a check result indicates a change in state." and, with respect to the observable behavior, this looks like a complete lie.

I've read [this] which seems to contradict the official documentation but gets it right:

> Event: An event is generated every time a check is done. It’s the result of the check, positive or negative, there’s always going to be one.

The solutions in that article are crazy for such a simple obvious task.

I've read [this blog post] and frankly I don't really get it. It seems to focus on performance, whereas I'm simply looking for reasonable behavior.

So... As of today, is there a (simple and clean) way to get notifications only on actual state-changes?

[documentation]: https://sensuapp.org/docs/0.25/reference/events.html#how-are-sensu-events-created
[this]: https://dmsimard.com/2016/03/21/tips-on-controlling-the-flow-of-notifications-in-sensu/
[this blog post]: https://sensuapp.org/blog/2016/07/07/sensu-plugin-filter-deprecation.html

**Comment by cwjohnston** [comment]
Hi @hchargois,

Thank you for raising this issue. You're absolutely correct; the documentation is lying. We'll fix that today.

I think the following [filter](https://sensuapp.org/docs/latest/reference/filters.html) would provide the behavior you are describing:

```json
{
  "filters": {
    "state_change_only": {
      "attributes": {
        "occurrences": "eval: value == 1 || :::action::: == :resolve"
      }
    }
  }
}
```

This filter works because the value of event `occurrences` is reset on state change, except when the event `action` is `:resolve`, e.g. as it is when returning status 0 after a previous non-zero status. You can find instructions for creating filter definitions and associating them with handlers here: https://sensuapp.org/docs/latest/reference/filters.html

Will you please try applying this filter and let me know if it provides the behavior you expect?

**Comment by cwjohnston** [comment]
I've updated the [documentation](https://sensuapp.org/docs/0.25/reference/events.html#how-are-sensu-events-created) to better describe how events are created. Thanks again for bringing this discrepancy to our attention.

**Comment by hchargois** [comment]
Thanks a lot for your speedy answer!

The documentation is much clearer now.

Your filter's logic is very straightforward, that's great. I tried it verbatim and it doesn't work, I have this in the server logs:

    {
      "timestamp": "2016-08-05T17:25:29.904855+0200",
      "level": "error",
      "message": "filter attribute eval error",
      "raw_eval_string": "eval: value == 1 || :::action::: == :resolve",
      "error": "undefined local variable or method `create' for Kernel:Module"
      [...]
    }

I guess that's because `:::action:::` gets replaced before evaluation and stands there without quotes or anything, so the ruby thinks it must be an identifier but doesn't know of it.

I tried:

    "eval: value == 1 || \":::action:::\" == \"resolve\""

And it works!

Thanks again!

**Comment by cwjohnston** [comment]
Hi @hchargois, thanks for reporting your success with the above and your feedback on the documentation change. 

Interesting that you needed to change the filter. In my test case the value of the event's `action` attribute comes through as a symbol, which is why I didn't need to quote the strings.

FWIW I tested the filter I described on Sensu 0.25.6, so if you are on an older version the filter may need to change when you upgrade.

I'd like to add this filter to our documentation as an example; can you help me by sharing which version of Sensu you are using with your modified filter?

**Comment by fessyfoo** [comment]
> I'd like to add this filter to our documentation as an example

👍

**Comment by fessyfoo** [comment]
quick doc feedback:  (worried I won't get back to  doing a PR or providing fuller response. ) 

the new docs currently say: 
> Sensu events are created to acknowledge that something potentially noteworthy has occurred

as well as:
> A Sensu Event is created every time a check result is processed by the Sensu server, regardless of the status indicated by the check result

this sounds contradictory , or that all check results are noteworthy.     and doesn't make it clear that green check results are generally filtered from the handlers (unless special conditions are involved such as the result being a metric )   so by default a handler for normal events with no configured filters sees failing results with status > 0 (`:create`, `:flapping`),  and the 1 successful resolution event with status 0  (:resolve)

hope that helps.

**Comment by hchargois** [comment]
> I'd like to add this filter to our documentation as an example; can you help me by sharing which version of Sensu you are using with your modified filter?

Weird... I'm also using 0.25.6 :confused:

**Comment by scosist** [comment]
I've tried both (just like your example @cwjohnston and placed in /etc/sensu/conf.d/filters.json)
```
"eval: value == 1 || :::action::: == :resolve"
```
and
```
"eval: value == 1 || \":::action:::\" == \"resolve\""
```
but still getting alerts every 30 minutes. Could there be something else I'm missing?

I'm using 0.25.6

**Comment by cwjohnston** [comment]
@scosist have you explicitly configured one or more handlers to use the filter?

**Comment by scosist** [comment]
Ah, I glossed over that part ([How do Sensu filters work?](https://sensuapp.org/docs/latest/reference/filters.html#how-do-sensu-filters-work)) of the filters reference. Thanks for the hint!
```
"filter": "state_change_only",
```
Looks like it's working now and I'm using the original example you provided--the `eval` without the extra escaped quotes. One thing to note, the filter(s) attribute must be defined on the handler and not the [handler set](https://sensuapp.org/docs/0.25/reference/handlers.html#handler-sets).

**Comment by cwjohnston** [comment]
Closing this as I've opened issues to track needed documentation improvements. Thanks for the feedback all!

**Comment by scosist** [comment]
I thought I should mention I experienced the same error that @hchargois did.

```
"error":"undefined local variable or method `create' for Kernel:Module"
```

This occurred after I switched from using the handler `filter` attribute to the `filters` attribute with more than one filter. Perhaps this has something to do with the check result containing `"action":"create"`? Also, once I started using more than one filter I stopped getting any alerts at all.

Filters:
```
    "state_change_only": {
      "negate": false,
      "attributes": {
        "occurrences": "eval: value == 1 || :::action::: == :resolve"
      }
    },
    "state_change_interval_60_daily": {
      "negate": false,
      "attributes": {
        "check": {
          "interval": 60
        },
        "occurrences": "eval: value == 1 || value % 1440 == 0 || :::action::: == :resolve"
      }
    }
```

Handler filters attribute:
```
      "filters": [
        "state_change_only",
        "state_change_interval_60_daily"
      ],
```

When I use @hchargois' workaround with the escaped double quotes, I no longer get the error but I'm still not getting any alerts.

Forgive me if I shouldn't comment on a closed issue; I believe this best applies here but if not, let me know and I'll create a new issue.

**Comment by scosist** [comment]
After further examination it seems I didn't test the original `eval` thoroughly enough (in the case of using one filter). I remember testing it to make sure it silenced subsequent handler alerts on the same state but I didn't test it for sending an `OK` alert on the `:resolve` state change. Instead it logs an error:
```
"error":"undefined local variable or method `resolve' for Kernel:Module"
```
Although sometimes instead of `method 'resolve'` the error is on `method 'create'`. Also, sometimes it _does_ send the `OK` alert but it doesn't seem to be reliable.

I found that using the following eval (with the action encased in single quotes) is sufficient (and more readable than escaping double quotes) for alerting on all state changes including `OK` and eliminates the logged error:
```
"eval: value == 1 || ':::action:::' == :resolve"
```
Unfortunately, this discovery does not resolve my issue of receiving zero alerts when using multiple filters.

**Comment by cwjohnston** [comment]
@scosist @hchargois in my testing I've learned that the only reason `"eval: value == 1 || :::action::: == :resolve"` (unquoted token and symbol comparison) appears to work is that the first half of the logic (`value == 1`) sometimes evaluates to true, which causes the rest of the condition to be skipped.

For any event where occurrences is not == 1, the filter will cause events to be dropped because of the `undefined local variable or method 'resolve' for Kernel:Module` error encountered when evaluating that second condition.

For future reference here is the working version using single quotes as described above:

```json
{
  "filters": {
    "state_change_only": {
      "attributes": {
        "occurrences": "eval: value == 1 || ':::action:::' == 'resolve'"
      }
    }
  }
}
```

**Comment by scosist** [comment]
This doesn't seem to alert on warning and critical states for checks with [something other](https://sensuapp.org/docs/latest/reference/plugins#check-definition-attributes) than the default of `1` for `occurrences`, for example `2`. I've tried a number of variations including:
```
"eval: value == :::check.occurrences|1::: || ':::action:::' == 'resolve'"
```
What would allow this filter to handle checks on state changes regardless of their occurrences value? Or is [this](https://github.com/sensu/sensu/issues/1175#issuecomment-198139861) not supported yet?

**Comment by scosist** [comment]
I still don't know whether the issue with the above example is a problem with the `check.occurrences` token or something else. I'm assuming that when a check definition has `"occurrences": 2` and the check result occurrences have reached `2` and an event is thus ready to be handled for the check, that the 'value' of `value` in the eval will also be `2` and thus return true for the `value == :::check.occurrences|1:::` statement, but if this were true I would expect it to alert on all state changes. Is it possible there is some confusion with the check definition occurrences attribute value and the check result occurrences value when using that token or is the [documentation ](https://sensuapp.org/docs/latest/reference/filters.html#example-filter-attribute-eval-token) correct? 

Either way, although I like it less, I think I've found a workaround. As long as I keep any of my `occurrences` [check definitions](https://sensuapp.org/docs/latest/reference/plugins#check-definition-attributes) to `5` or less the following filter seems to work for alerting on all state changes:
```
{
  "filters": {
    "state_change_only": {
      "negate": true,
      "attributes": {
        "occurrences": "eval: value > 5 && ':::action:::' != 'resolve'"
      }
    }
  }
}
```

If anyone has any insights as to the problem above, please share.

@cwjohnston any chance we could reopen this? It seems the most recent example you provided only works entirely as expected when using the default value of `1` for `occurrences` on all check definitions.

**Comment by ctal80** [comment]
HI, 

I'm trying to use the following filter: 

"occurrences": "eval: value == 1 || ':::action:::' == 'resolve'"


which seems to filter some of the resolved events. 

and advice what can be done in order to get notifications only on state change properly?

Thanks 

Tal

**Comment by mcbsdEU** [comment]
@ctal80 You did not provide any details...just generic statement.

**Comment by ctal80** [comment]
following is an example where we can see that the event was filtered although action":"resolve: 

{"timestamp":"2017-03-21T12:51:25.237159+0200","level":"info","message":"event was filtered","handler":{"type":"pipe","command":"/opt/sensu/embedded/bin/ruby /etc/sensu/plugins/handler-slack-custom.rb","filter":"state_change_only","name":"slack"},"event":{"id":"cf86d95e-2a65-48da-b7cd-7bfc0116943d","client":{"name":"P4_proxy_AMR","address":"unknown","subscriptions":[],"keepalives":false,"version":"0.20.3"},"check":{"command":"powershell.exe -ExecutionPolicy ByPass -f /opt/sensu/plugins/test.ps1 p4 \" ","interval":3600,"subscribers":["roundrobin:web_check"],"handlers":["mailer","slack"],"source":"P4_proxy_AMR","name":"P4_Proxy_AMR_Check","issued":1490093482,"executed":1490093482,"duration":2.804,"output":"837:45:02\r\nCheck OK: \n","status":0,"history":["1","1","1","1","1","1","1","1","0","0","1","0","1","1","0","1","1","1","1","1","0"],"total_state_change":37},"occurrences":5,"action":"resolve","timestamp":1490093485}}
